### PR TITLE
perf(core): make more `ApplicationModule` providers tree-shakable

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This PR refactors `ApplicationModule` providers to make them tree-shakable. See individual commits for additional details.

## PR Type
What kind of change does this PR introduce?

- [x] Performance improvement

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/application_module.ts [hunk 1]
```diff
@@ -31,85 +19,15 @@ export function _keyValueDiffersFactory() {
...
-  },
-  {provide: SCHEDULER, deps: [NgZone], useFactory: zoneSchedulerFactory},
-  {
-    provide: ApplicationInitStatus,
-    useClass: ApplicationInitStatus,
-    deps: [[new Optional(), APP_INITIALIZER]]
-  },
-  {provide: Compiler, useClass: Compiler, deps: []},
-  APP_ID_RANDOM_PROVIDER,
   {provide: IterableDiffers, useFactory: _iterableDiffersFactory, deps: []},
```
Comment (AndrewKushnir) [inline packages/core/src/application_module.ts:27]: Note: the `IterableDiffers` and `KeyValueDiffers` are refactored in PR #45094.

**Comment by AndrewKushnir** [comment]
[Presubmit](https://test.corp.google.com/OCL:428961434:BASE:428961528:1644994332853:12e8d80e).

**Review by atscott — APPROVED** [review]

**Review by atscott — APPROVED** [review]
reviewed-for: size-tracking

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

reviewed-for: size-tracking

**Comment by AndrewKushnir** [comment]
["Green" TGP](https://test.corp.google.com/ui#id=OCL:428961434:BASE:429069945:1645033181963:f91ac76b).

**Comment by alxhub** [comment]
Note: this doesn't merge cleanly to `13.2.x`.

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 1e60fe0a3eeb6bd46997f7eeb4825351cfbe3c44.

**Pull request closed without merging**